torch/_decomp/decompositions_for_rng.py

```diff
@@ -0,0 +1,292 @@
+    shape.
+    """
+    stride = [1] * len(size) + [1]
+    for idx in reversed(range(len(size))):
+        stride[idx] = stride[idx + 1] * size[idx]
+    stride = stride[1:]
+    return stride
+
+
+class RNGFunctionalizationError(Exception):
+    pass
+
+
+def throw_on_cpu():
+    raise RNGFunctionalizationError(
+        "You are trying to functionalize a CPU RNG operator but CPU does not use "
+        "Philox/counter-based PRNG. Therefore, functionalizing a CPU RNG operator "
+        "is not supported. We are discussing the possibility of a Philox-based "
+        "RNG implementation for CPU. We will revisit this assertion in future."
+    )
+
+
+@register_offset_calculation([aten.rand])
+def rand_offset_calculator(shape):
+    # For impl, look at calc_execution_policy
+    numel = 1
+    for dim_size in shape:
+        numel *= dim_size
+
+    block_size = 256
+    unroll = 4
+    curand4_engine_calls = 4
+    device_property = torch.cuda.get_device_properties(torch.cuda.current_device())
+    blocks_per_sm = int(device_property.max_threads_per_multi_processor / block_size)
+    grid_size = int((numel + block_size - 1) / block_size)
+    grid_size = min(grid_size, device_property.multi_processor_count * blocks_per_sm)
+    offset = (
+        int((numel - 1) / (block_size * grid_size * unroll) + 1) * curand4_engine_calls
+    )
+    return offset
+
+
+# TODO - We have to register many more distributions here, and also higher level
+# ops like dropout which have fused implementation and can hide the rand inside.
+@register_rng_decomposition(aten.rand)
+def rand(shape, dtype=None, layout=torch.strided, device=None, pin_memory=False):
+    device = device or "cpu"
+
+    if device.type == "cpu":
+        throw_on_cpu()
+    seed, offset = PhiloxStateTracker.get_state_as_tuple()
+    dtype = dtype or torch.float32
+    stride = get_default_stride(shape)
+    r = torch.ops.prims.philox_rand(shape, seed, offset, stride, device, dtype)
+    PhiloxStateTracker.advance_offset(offset_calculator[aten.rand](shape))
+    return r
+
+
+@register_rng_decomposition(aten.rand_like)
+def rand_like(
+    x: torch.Tensor,
+    dtype=None,
+    layout=None,
+    device=None,
+    pin_memory=False,
+    memory_format=torch.preserve_format,
+):
+    device = device or x.device
+    if device.type == "cpu":
+        throw_on_cpu()
+    dtype = dtype or x.dtype
+    seed, offset = PhiloxStateTracker.get_state_as_tuple()
+    r = torch.ops.prims.philox_rand(x.shape, seed, offset, x.stride(), device, dtype)
+    PhiloxStateTracker.advance_offset(offset_calculator[aten.rand](x.shape))
+    return r
+
+
+class PhiloxState:
+    """
+    Represents a PhiloxRngState - (seed, offset) where offset = base_offset +
+    relative_offset. seed and base_offset basically point to the rng state just
+    before tracing starts. relative offset tracks the totaly consumed offset at
+    trace time.
+    """
+
+    def __init__(self):
+        self.reset()
+
+    def reset(self):
```

Comment:
`reset` is used to do a clean slate, so that next time AOTAutograd is called on a new graph, we don't pollute.

This is required because `PhiloxStateTracker` is a singleton object. It is required at multiple places - like AOT Autograd and inside rng decompositions. So, `reset` is called to do a clean slate. I am adding, hopefully a clearer, method - `mark_end_of_tracing` which calls `reset`

Reply:
But after the reset is called, this cannot be used as is right? We need to put the proper seed/original offset before you can start tracking stuff right?

Reply:
Yes,  so we set all of this up in `create_functionalized_rng_ops_wrapper`

The main piece of code is 
![image](https://user-images.githubusercontent.com/13822661/230997070-2d3f1fb5-98f9-453a-90ac-1f8e44fe16ba.png)

Before we start tracing, we set the fwd and bwd state.

Reply:
Right so reset should not look like a valid state. it should raise an error if you try to update it before you set the original state